lib/rubygems/source.rb

```diff
@@ -194,7 +194,15 @@ def load_specs(type)
     spec_dump = fetcher.cache_update_path spec_path, local_file, update_cache?
 
     begin
-      Gem::NameTuple.from_list Marshal.load(spec_dump)
+      gem_name_tuples = Gem::NameTuple.from_list Marshal.load(spec_dump)
+
+      if type == :prerelease && is_uri_for_github_rubygems?(uri) && gem_name_tuples.length > 0 && gem_name_tuples.all? {|t| t.version == Gem::Version.new(0) }
```

Comment:
I think it would be better to capture the if condition into a private function. I'll do it after I got the ✅ .

Reply:
Let's go through the conditions:

1. That GitHub issue is only happening for the `prerelease.x.x.gz` file so I thought it would be good to have it as a condition here.
2. making sure the URL is for the GitHub RubyGems server.
3. make sure at least one gem tuple name has been generated
4. make sure at least one gem tuple with version 0 exist

Technically, using version `0.0.0` is not against the Semantic Versioning so I didn't want to blindly warn users for using the `0` or `0.0.0` version. [It's discussed here too.](https://github.com/semver/semver/issues/363)
Also, a user could actually create a `0.0.0` gem in the GitHub Rubygems repository so I thought it wouldn't be a good idea to raise an exception to break the process.

Reply:
I see.

My initial idea was meant to be completely independent from GitHub, so not require any hard-coded checks like the ones here.

Initially I thought of warning this case, but then I realized we might not need to warn this case at all, but instead avoid fetching this invalid URI at all, since it's unneeded.

These invalid URIs correspond to `0.0.0` versions which, even if valid as you point out, it should always take the lowest precedence, so as long as we can find other gems with higher version in the index that don't have this problem, we shouldn't have to resort to fetching this URL at all.

From looking at the backtrace of the error, it comes from here:

https://github.com/rubygems/rubygems/blob/c693b899c48197b6411dfac43e80f2c36eb6961a/lib/rubygems/resolver/installer_set.rb#L80-L96

I think, if you change that `found.select` to `found.reverse.find`, then you avoid fetching all potential specs, and instead fetch them one by one in order of priority. If you make this change, you need to also avoid the surrounding code to catch up with the fact that we're now expecting either `nil` or a spec, rather than either an empty array or a non-empty array of specs.

It's completely untested, but hopefully you get the idea and can verify whether it works.

Reply:
@intuxicated I didn't verify whether it fixes the issue since I didn't setup an environment to reproduce it, but I did check that my suggestion breaks no existing specs, which is something. The specific diff would be

```diff
diff --git a/lib/rubygems/resolver/installer_set.rb b/lib/rubygems/resolver/installer_set.rb
index 60181315b0..f4fee351a5 100644
--- a/lib/rubygems/resolver/installer_set.rb
+++ b/lib/rubygems/resolver/installer_set.rb
@@ -77,11 +77,11 @@ def add_always_install(dependency)
     newest = found.last
 
     unless @force
-      found_matching_metadata = found.select do |spec|
+      found_matching_metadata = found.reverse.find do |spec|
         metadata_satisfied?(spec)
       end
 
-      if found_matching_metadata.empty?
+      if found_matching_metadata.nil?
         if newest
           ensure_required_ruby_version_met(newest.spec)
           ensure_required_rubygems_version_met(newest.spec)
@@ -92,7 +92,7 @@ def add_always_install(dependency)
           raise exc
         end
       else
-        newest = found_matching_metadata.last
+        newest = found_matching_metadata
       end
     end
```

Reply:
Thanks @deivid-rodriguez

Reply:
It worked!